### PR TITLE
Ik/ote demo

### PR DESCRIPTION
**Pull request by Ilya-Krylov**

### CODE REVIEW

**Comment by Ilya-Krylov** [comment]
run ote_sdk tests

**Comment by Ilya-Krylov** [comment]
run ote_sdk test2

**Comment by Ilya-Krylov** [comment]
run ote_sdk tests2

**Comment by Ilya-Krylov** [comment]
run ote_sdk tests2

**Comment by Ilya-Krylov** [comment]
run ote_sdk tests2

**Comment by Ilya-Krylov** [comment]
run ote_sdk tests

**Comment by Ilya-Krylov** [comment]
run ote_sdk tests

File: ote_cli/ote_cli/tools/demo.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
+"""
+Model inference demostration tool.
```
Comment (druzhkov-paul) [inline ote_cli/ote_cli/tools/demo.py:2]: demon

File: ote_cli/ote_cli/tools/demo.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    hyper_parameters = template.hyper_parameters.data
+    assert hyper_parameters
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("template")
+    parser.add_argument(
+        "-i",
+        "--input",
+        required=True,
+        help="Source of input data: images folder, image, webcam.",
```
Comment (druzhkov-paul) [inline ote_cli/ote_cli/tools/demo.py:70]: video?

File: ote_cli/ote_cli/tools/demo.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        "This options applies to result visualisation only.",
+    )
+    parser.add_argument(
+        "--loop", action="store_true", help="Enable reading the input in a loop."
+    )
+    parser.add_argument(
+        "--delay", type=int, default=3, help="Frame visualization time in ms."
+    )
+    parser.add_argument(
+        "--labels", nargs="+", required=True, help="A space-separated labels list."
```
Comment (druzhkov-paul) [inline ote_cli/ote_cli/tools/demo.py:92]: IMO, it would be good to describe all required arguments first and then the rest.

File: ote_cli/ote_cli/tools/demo.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    dataset = DatasetEntity(items=[item])
+
+    start_time = time.perf_counter()
+    predicted_validation_dataset = task.infer(
+        dataset,
+        InferenceParameters(is_evaluation=True),
+    )
+    elapsed_time = time.perf_counter() - start_time
+    item = predicted_validation_dataset[0]
+    return item.annotation_scene.annotations, elapsed_time
```
Comment (druzhkov-paul) [inline ote_cli/ote_cli/tools/demo.py:130]: It's better use `item.get_annotations()` rather than get those directly from the annotation scene. I guess the result will be the same in this case (all ROIs are covering the whole frame), but in general the recommendation is to do so.

File: ote_cli/ote_cli/tools/demo.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+
+def main():
+    """
+    Main function that is used for model evaluation.
```
Comment (druzhkov-paul) [inline ote_cli/ote_cli/tools/demo.py:135]: It's not evaluation.

File: ote_cli/ote_cli/tools/demo.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    """
+
+    # Dynamically create an argument parser based on override parameters.
+    args, template, hyper_parameters = parse_args()
+    # Get new values from user's input.
+    updated_hyper_parameters = gen_params_dict_from_args(args)
+    # Override overridden parameters by user's values.
```
Comment (druzhkov-paul) [inline ote_cli/ote_cli/tools/demo.py:142]: Let's go deeper with this. 😄

File: ote_cli/ote_cli/tools/demo.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    override_parameters(updated_hyper_parameters, hyper_parameters)
+
+    hyper_parameters = create(hyper_parameters)
+
+    labels = [
+        LabelEntity(l, template.task_type, id=ID(i)) for i, l in enumerate(args.labels)
+    ]
+
+    # Get classes for Task, ConfigurableParameters and Dataset.
+    taks_class = get_impl_class(template.entrypoints.base)
```
Comment (druzhkov-paul) [inline ote_cli/ote_cli/tools/demo.py:152]: task

File: ote_cli/ote_cli/tools/demo.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+        model_adapters={
+            "weights.pth": ModelAdapter(load_model_weights(args.load_weights))
+        },
+        train_dataset=None,
+    )
+    environment.model = model
+
+    task = taks_class(task_environment=environment)
+
+    capture = open_images_capture(args.input, args.loop)
```
Comment (druzhkov-paul) [inline ote_cli/ote_cli/tools/demo.py:171]: Like [Open Images](https://opensource.google/projects/open-images-dataset)?

File: ote_cli/ote_cli/tools/demo.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    frame_index = 0
+    while True:
+        frame = capture.read()
+        if frame is None:
+            break
+
+        predictions, elapsed_time = get_predictions(task, frame)
+        elapsed_times.append(elapsed_time)
+        if len(elapsed_times) > 10:
+            elapsed_times = elapsed_times[1:]
```
Comment (druzhkov-paul) [inline ote_cli/ote_cli/tools/demo.py:183]: You may consider using `collections.deque` instead with `maxlen` specified.

File: ote_cli/ote_cli/tools/demo.py [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+        elapsed_time = np.mean(elapsed_times)
+
+        frame = draw_predictions(
+            template.task_type, predictions, frame, args.fit_to_size
+        )
```
Comment (druzhkov-paul) [inline ote_cli/ote_cli/tools/demo.py:188]: I guess visualization can be places in the following `if` clause.

File: ote_cli/ote_cli/tools/utils/demo/visualization.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    )
+    # Blend original image with the one, where instances are colored.
+    # As a result instances masks become transparent.
+    cv2.addWeighted(frame, 0.5, segments_image, 0.5, 0, dst=frame)
+    return frame
+
+
+def put_labels(frame, predictions):
+    """
+    Coverts predictions to text labels and puts them to the top left corner of a frame.
```
Comment (druzhkov-paul) [inline ote_cli/ote_cli/tools/utils/demo/visualization.py:101]:
```suggestion
    Converts predictions to text labels and puts them to the top left corner of a frame.
```

File: ote_cli/ote_cli/tools/demo.py (hunk 3 above)
Comment (druzhkov-paul) [reply ote_cli/ote_cli/tools/demo.py:92]: And what about saving/loading the labels as a part persistent model representation?

Comment (Ilya-Krylov) [reply ote_cli/ote_cli/tools/demo.py:92]: oops

File: ote_cli/ote_cli/tools/demo.py (hunk 8 above)
Comment (Ilya-Krylov) [reply ote_cli/ote_cli/tools/demo.py:171]: ho-ho

File: ote_cli/ote_cli/tools/demo.py (hunk 1 above)
Comment (Ilya-Krylov) [reply ote_cli/ote_cli/tools/demo.py]: fixed

File: ote_cli/ote_cli/tools/demo.py (hunk 2 above)
Comment (Ilya-Krylov) [reply ote_cli/ote_cli/tools/demo.py]: and video

File: ote_cli/ote_cli/tools/demo.py (hunk 4 above)
Comment (Ilya-Krylov) [reply ote_cli/ote_cli/tools/demo.py]: fixed

File: ote_cli/ote_cli/tools/demo.py (hunk 5 above)
Comment (Ilya-Krylov) [reply ote_cli/ote_cli/tools/demo.py]: fixed

File: ote_cli/ote_cli/tools/demo.py (hunk 9 above)
Comment (Ilya-Krylov) [reply ote_cli/ote_cli/tools/demo.py]: fixed

File: ote_cli/ote_cli/tools/demo.py (hunk 10 above)
Comment (Ilya-Krylov) [reply ote_cli/ote_cli/tools/demo.py]: fixed

File: ote_cli/ote_cli/tools/demo.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+    labels = [
+        LabelEntity(l, template.task_type, id=ID(i)) for i, l in enumerate(args.labels)
+    ]
+
+    # Get classes for Task, ConfigurableParameters and Dataset.
+    task_class = get_impl_class(template.entrypoints.base)
+    environment = TaskEnvironment(
+        model=None,
+        hyper_parameters=hyper_parameters,
+        label_schema=LabelSchemaEntity.from_labels(labels),
```
Comment (sovrasov) [inline ote_cli/ote_cli/tools/demo.py:157]: This wouldn't work for multilabel: classification makes a decision which config to used based in the label schema. To add multilabel classification this step should be customized. The same is true for hierarchical classification.

**Review by sovrasov — APPROVED** [review]

**Review by druzhkov-paul — APPROVED** [review]

File: ote_cli/ote_cli/tools/demo.py (hunk 12 above)
Comment (Ilya-Krylov) [reply ote_cli/ote_cli/tools/demo.py:157]: Will work on this issue and Reading/Writing labels from/to snapshot in next PR(s)

Comment (Ilya-Krylov) [reply ote_cli/ote_cli/tools/demo.py:157]: @druzhkov-paul

**Comment by Ilya-Krylov** [comment]
run ote_sdk tests
